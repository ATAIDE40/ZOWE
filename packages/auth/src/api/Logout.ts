/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { AbstractSession, ImperativeExpect, Logger, ImperativeError, HTTP_VERB, RestConstants, SessConstants } from "@zowe/imperative";
import { ZosmfRestClient } from "../../../rest";

/**
 * Class to handle logging out of APIML.
 * @export
 * @class Logout
 */
export class Logout {

    /**
     * Perform APIML logout to invalidate LTPA2 or other token types.
     * @static
     * @param {AbstractSession} session
     * @returns
     * @memberof Login
     */
    public static async apimlLogout(session: AbstractSession, request: HTTP_VERB, resource: string) {
        Logger.getAppLogger().trace("Logout.logout()");
        ImperativeExpect.toNotBeNullOrUndefined(session, "Required session must be defined");

        const client = new ZosmfRestClient(session);
        await client.request({
            request,
            resource
        });

        ImperativeExpect.toNotBeNullOrUndefined(session.ISession.tokenValue, "Session token not populated. Unable to logout.");

        if (client.response.statusCode !== RestConstants.HTTP_STATUS_204) {
            throw (client as any).populateError({
                msg: `REST API Failure with HTTP(S) Status ${client.response.statusCode}`,
                source: SessConstants.HTTP_PROTOCOL
            });
        }
    }
}
