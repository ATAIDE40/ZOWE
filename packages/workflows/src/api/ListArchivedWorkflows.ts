/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ZosmfRestClient } from "../../../rest";
import { WorkflowValidator } from "./WorkflowValidator";
import { AbstractSession, Headers } from "@brightside/imperative";
import { WorkflowConstants, nozOSMFVersion } from "./WorkflowConstants";
import { IWorkflowInfo } from "./doc/IWorkflowInfo";
import { IActiveWorkflows } from "./doc/IActiveWorkflows";


export class ListArchivedWorkflows {
   public static async listArchivedWorkflows(session: AbstractSession, workflowKey?: string,
                                             zOSMFVersion = WorkflowConstants.ZOSMF_VERSION): Promise<IActiveWorkflows> {
        WorkflowValidator.validateSession(session);
        WorkflowValidator.validateNotEmptyString(zOSMFVersion, nozOSMFVersion.message);
        let resourcesQuery: string = `${WorkflowConstants.RESOURCE}/${zOSMFVersion}/`;
        if (workflowKey){
            resourcesQuery += `${WorkflowConstants.ARCH_WORKFLOW_RESOURCE}/${workflowKey}`;
        }
         else  {
        resourcesQuery += `${WorkflowConstants.ARCH_WORKFLOW_RESOURCE}`;
              }
        return ZosmfRestClient.getExpectJSON (session, resourcesQuery, [Headers.APPLICATION_JSON]);
    }
}
