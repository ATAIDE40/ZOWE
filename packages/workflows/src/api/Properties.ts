/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/


import { AbstractSession, Headers } from "@brightside/imperative";
import { ZosmfRestClient } from "../../../rest";
import { WorkflowConstants, nozOSMFVersion,
        noOwner, noWorkflowKey } from "./WorkflowConstants";
import { WorkflowValidator } from "./WorkflowValidator";
import { isNullOrUndefined } from "util";
import { IWorkflowInfo } from "./doc/IWorkflowInfo";
import { IStepSummary } from "./doc/IStepSummary";
import { IStepInfo } from "./doc/IStepInfo";

export class PropertiesWorkflow {
    /**
     * This operation returns properties of the workflow.
     * Parameters indicators are mandatory,request can include steps and variables indicator for requested result.
     * @param {AbstractSession} session - z/OSMF connection info
     * @param {string} workflowfKey - Key of workflow.
     * @param {string} zOSMFVersion - the URI path that identifies the version of the provisioning service.
     * @param {boolean} steps - Optional parameter for listing steps properties.
     * @param {boolean} variables - Optional parameter for listing variables properties.
     * @returns {Promise<IWorkflowInfo>} z/OSMF response object
     * @memberof Properties
     */
    // main method
    public static async getWorkflowProperties(session: AbstractSession, workflowKey: string, zOSMFVersion = WorkflowConstants.ZOSMF_VERSION,
                                              steps?: boolean, variables?: boolean): Promise<IWorkflowInfo>{
        WorkflowValidator.validateSession(session);
        WorkflowValidator.validateNotEmptyString(zOSMFVersion, nozOSMFVersion.message);
        let wfKey: string;

        WorkflowValidator.validateNotEmptyString(workflowKey, noWorkflowKey.message);
        wfKey = workflowKey;

        let resourcesQuery: string = `${WorkflowConstants.RESOURCE}/${zOSMFVersion}/`;
        resourcesQuery += `${WorkflowConstants.WORKFLOW_RESOURCE}/${wfKey}`;

        if (steps && variables){
            resourcesQuery += `?${WorkflowConstants.returnData}=${WorkflowConstants.steps},${WorkflowConstants.variables}`;

        } else if (steps)   {
            resourcesQuery += `?${WorkflowConstants.returnData}=${WorkflowConstants.steps}`;

        } else if (variables)   {
            resourcesQuery += `?${WorkflowConstants.returnData}=${WorkflowConstants.variables}`;
        }

        return ZosmfRestClient.getExpectJSON<IWorkflowInfo>(session, resourcesQuery, [Headers.APPLICATION_JSON]);
    }

    /**
     * Returns the summary of the steps only
     *
     * @static
     * @param {AbstractSession} session - z/OSMF connection info
     * @param {string} workflowKey - Key of workflow.
     * @param {string} [zOSMFVersion=WorkflowConstants.ZOSMF_VERSION] - the URI path that identifies the version of the provisioning service.
     * @returns {Promise<IStepSummary[]>} z/OSMF response object
     * @memberof PropertiesWorkflow
     */
    public static async getSummaryOnly(session: AbstractSession, workflowKey: string,
                                       zOSMFVersion = WorkflowConstants.ZOSMF_VERSION): Promise<IStepSummary[]> {
        WorkflowValidator.validateSession(session);
        WorkflowValidator.validateNotEmptyString(workflowKey, noWorkflowKey.message);

        let resourcesQuery: string = `${WorkflowConstants.RESOURCE}/${zOSMFVersion}/`;
        resourcesQuery += `${WorkflowConstants.WORKFLOW_RESOURCE}/${workflowKey}`;
        resourcesQuery += `?${WorkflowConstants.returnData}=${WorkflowConstants.steps}`;

        const response: IWorkflowInfo = await ZosmfRestClient.getExpectJSON(session, resourcesQuery, [Headers.APPLICATION_JSON]);
        const steps: IStepInfo[] = response.steps;
        let stepSummaries: IStepSummary[] = [];

        stepSummaries = await PropertiesWorkflow.processStepSummaries(steps);

        return stepSummaries;
    }

    /**
     * Processes the z/OSMF workflow step info
     * in a recursive manner.
     *
     * @protected
     * @static
     * @param {IStepInfo[]} steps z/OSMF steps to be processed
     * @returns {Promise<IStepSummary[]>} Array of z/OSMF step summary objects
     * @memberof PropertiesWorkflow
     */
    protected static async processStepSummaries(steps: IStepInfo[]): Promise<IStepSummary[]> {
        let stepSummaries: IStepSummary[] = [];

        for(const step of steps) {
            let miscValue: string = "N/A";
            if(step.submitAs && step.submitAs.match(/.*JCL/)) {
                if(step.jobInfo && step.jobInfo.jobstatus) {
                    miscValue = step.jobInfo.jobstatus.jobid;
                }
            } else if(step.template) {
                miscValue = "TSO";
            } else if(step.isRestStep) {
                miscValue = step.actualStatusCode;
            }
            const stepSummary: IStepSummary = {
                id: step.stepNumber,
                name: step.name,
                state: step.state,
                misc: miscValue
            };

            stepSummaries.push(stepSummary);
            if(step.steps) {
                const subSteps = await PropertiesWorkflow.processStepSummaries(step.steps);
                stepSummaries = stepSummaries.concat(subSteps);
            }
        }

        return stepSummaries;
    }
}

