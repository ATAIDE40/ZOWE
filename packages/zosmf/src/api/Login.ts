
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
import { AbstractSession, ImperativeExpect, Logger, ImperativeError } from "@zowe/imperative";
import { ZosmfRestClient } from "../../../rest";
import { LoginConstants } from "./LoginConstants";

/**
 * Class to handle logging onto z/OSMF.
 * @export
 * @class Login
 */
export class Login {

    /**
     * Perform z/OSMF login to obtain LTPA2 or other token types.
     * @static
     * @param {AbstractSession} session
     * @returns
     * @memberof Login
     */
    public static async login(session: AbstractSession) {
        Logger.getAppLogger().trace("Login.login()");
        ImperativeExpect.toNotBeNullOrUndefined(session, "Required session must be defined");
        await ZosmfRestClient.getExpectJSON<any>(session, LoginConstants.RESOURCE); // TODO(Kelosky): eventually a real auth endpoint
        return session.ISession.tokenValue;
    }

}
