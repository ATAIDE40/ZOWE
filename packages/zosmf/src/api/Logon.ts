
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
import { AbstractSession, ImperativeExpect, Logger, ImperativeError } from "@zowe/imperative";
import { ZosmfRestClient } from "../../../rest";
import { LogonConstants } from "./LogonConstants";

/**
 * Class to handle logging onto z/OSMF.
 * @export
 * @class Logon
 */
export class Logon {

    /**
     * Perform z/OSMF logon to obtain LTPA2 or other token types.
     * @static
     * @param {AbstractSession} session
     * @returns
     * @memberof Logon
     */
    public static async logon(session: AbstractSession) {
        Logger.getAppLogger().trace("Logon.logon()");
        ImperativeExpect.toNotBeNullOrUndefined(session, "Required session must be defined");
        await ZosmfRestClient.getExpectJSON<any>(session, LogonConstants.RESOURCE); // TODO(Kelosky): eventually a real auth endpoint
        return session.ISession.tokenValue;
    }

}
