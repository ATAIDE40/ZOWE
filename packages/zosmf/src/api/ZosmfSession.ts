/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ICommandArguments, ICommandOptionDefinition, IProfile, Logger, Session, ISession, ImperativeError } from "@zowe/imperative";

/**
 * Utility Methods for Brightside
 * @export
 */
export class ZosmfSession {

    public static ZOSMF_CONNECTION_OPTION_GROUP = "Zosmf Connection Options";

    /**
     * Option used in profile creation and commands for hostname for z/OSMF
     */
    public static ZOSMF_OPTION_HOST: ICommandOptionDefinition = {
        name: "host",
        aliases: ["H"],
        description: "The z/OSMF server host name.",
        type: "string",
        group: ZosmfSession.ZOSMF_CONNECTION_OPTION_GROUP
    };

    /**
     * Option used in profile creation and commands for port for z/OSMF
     */
    public static ZOSMF_OPTION_PORT: ICommandOptionDefinition = {
        name: "port",
        aliases: ["P"],
        description: "The z/OSMF server port.",
        type: "number",
        defaultValue: 443,
        group: ZosmfSession.ZOSMF_CONNECTION_OPTION_GROUP
    };

    /**
     * Option used in profile creation and commands for username / ID  for z/OSMF
     */
    public static ZOSMF_OPTION_USER: ICommandOptionDefinition = {
        name: "user",
        aliases: ["u"],
        description: "Mainframe (z/OSMF) user name, which can be the same as your TSO login.",
        type: "string",
        group: ZosmfSession.ZOSMF_CONNECTION_OPTION_GROUP
    };

    /**
     * Option used in profile creation and commands for password/passphrase for z/OSMF
     */
    public static ZOSMF_OPTION_PASSWORD: ICommandOptionDefinition = {
        name: "password",
        aliases: ["pass", "pw"],
        description: "Mainframe (z/OSMF) password, which can be the same as your TSO password.",
        type: "string",
        group: ZosmfSession.ZOSMF_CONNECTION_OPTION_GROUP
    };

    /**
     * Option used in profile creation and commands for rejectUnauthorized setting for connecting to z/OSMF
     */
    public static ZOSMF_OPTION_REJECT_UNAUTHORIZED: ICommandOptionDefinition = {
        name: "reject-unauthorized",
        aliases: ["ru"],
        description: "Reject self-signed certificates.",
        type: "boolean",
        defaultValue: true,
        group: ZosmfSession.ZOSMF_CONNECTION_OPTION_GROUP
    };

    /**
     * Option used in profile creation and commands for base path setting for connecting to z/OSMF
     */
    public static ZOSMF_OPTION_BASE_PATH: ICommandOptionDefinition = {
        name: "base-path",
        aliases: ["bp"],
        description: "The base path for your API mediation layer instance." +
            " Specify this option to prepend the base path to all z/OSMF resources when making REST requests." +
            " Do not specify this option if you are not using an API mediation layer.",
        type: "string",
        group: ZosmfSession.ZOSMF_CONNECTION_OPTION_GROUP
    };

    /**
     * Option used in profile creation and commands to indicate token-based connection to z/OSMF
     */
    public static ZOSMF_OPTION_TOKEN_TYPE: ICommandOptionDefinition = {
        name: "token-type",
        aliases: ["tt"],
        description: "The token type for z/OSMF returned in the HTTP Cookie header, for example: 'LtpaToken2'." +
            " The existence of a 'tokenType' indicates that basic authentication will not be used after a token has been obtained.",
        type: "stringOrEmpty",
        group: ZosmfSession.ZOSMF_CONNECTION_OPTION_GROUP
    };

    /**
     * Option used in profile creation and commands to indicate the token value for connecting to z/OSMF
     */
    public static ZOSMF_OPTION_TOKEN_VALUE: ICommandOptionDefinition = {
        name: "token-value",
        aliases: ["tv"],
        description: "The token value associated with the token type.",
        type: "string",
        group: ZosmfSession.ZOSMF_CONNECTION_OPTION_GROUP
    };

    /**
     * Options related to connecting to z/OSMF
     * These options can be filled in if the user creates a profile
     */
    public static ZOSMF_CONNECTION_OPTIONS: ICommandOptionDefinition[] = [
        ZosmfSession.ZOSMF_OPTION_HOST,
        ZosmfSession.ZOSMF_OPTION_PORT,
        ZosmfSession.ZOSMF_OPTION_USER,
        ZosmfSession.ZOSMF_OPTION_PASSWORD,
        ZosmfSession.ZOSMF_OPTION_REJECT_UNAUTHORIZED,
        ZosmfSession.ZOSMF_OPTION_BASE_PATH,
        ZosmfSession.ZOSMF_OPTION_TOKEN_TYPE,
        ZosmfSession.ZOSMF_OPTION_TOKEN_VALUE
    ];


    /**
     * Given a z/OSMF profile, create a REST Client Session.
     * @static
     * @param {IProfile} profile - The z/OSMF profile contents
     * @returns {Session} - A session for usage in the z/OSMF REST Client
     */
    public static createBasicZosmfSession(profile: IProfile): Session {
        this.log.debug("Creating a z/OSMF session from the profile named %s", profile.name);
        return new Session({
            type: "basic",
            hostname: profile.host,
            port: profile.port,
            user: profile.user,
            password: profile.password,
            rejectUnauthorized: profile.rejectUnauthorized,
            basePath: profile.basePath
        });
    }

    /**
     * Given command line arguments, create a REST Client Session.
     * @static
     * @param {IProfile} args - The arguments specified by the user
     * @returns {Session} - A session for usage in the z/OSMF REST Client
     */
    public static createBasicZosmfSessionFromArguments(args: ICommandArguments): Session {
        this.log.debug("Creating a z/OSMF session from arguments");

        const sessionConfig: ISession = {
            hostname: args.host,
            port: args.port,
            rejectUnauthorized: args.rejectUnauthorized,
            basePath: args.basePath
        };

        this.log.debug("Using basic authentication");
        sessionConfig.type = "basic";
        sessionConfig.user = args.user;
        sessionConfig.password = args.password;

        if (args.tokenType && args.tokenValue && args.tokenValue !== "") {
            this.log.debug("Using token authentication");
            sessionConfig.type = "token";
            sessionConfig.tokenType = args.tokenType;
            sessionConfig.tokenValue = args.tokenValue;
        }

        return new Session(sessionConfig);
    }


    private static get log(): Logger {
        return Logger.getAppLogger();
    }
}
