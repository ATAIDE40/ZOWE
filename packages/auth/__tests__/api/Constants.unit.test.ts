/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { LoginConstants } from "../../src/api/LoginConstants";
import { LogoutConstants } from "../../src/api/LogoutConstants";

describe("LoginConstants Unit Test", () => {
    it("Should not have changed", () => {
        expect(LoginConstants).toMatchSnapshot();
    });
});

describe("LogoutConstants Unit Test", () => {
    it("Should not have changed", () => {
        expect(LogoutConstants).toMatchSnapshot();
    });
});
