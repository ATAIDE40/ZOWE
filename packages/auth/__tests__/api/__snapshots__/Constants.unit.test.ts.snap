// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginConstants Unit Test Should not have changed 1`] = `[Function]`;

exports[`LogoutConstants Unit Test Should not have changed 1`] = `[Function]`;
